File: 22-generate-parentheses/generate-parentheses.java
class Solution {
    
    public List<String> generateParenthesis(int n) {
        Set<String> ret = new HashSet<>();
        ret.add("()");
        ret = helper(2, ret, n);
        return new ArrayList<>(ret);
    
    }

    public Set<String> helper(int n, Set<String> ret, int give)
    {
        if(n == give+1){
            return ret;
        }
        Set<String> temp = new HashSet<>();
        for(String st : ret){
            temp.add("()" + st);
            temp.add(st+"()");
            // temp.add('('+st+')');
            int ind = st.indexOf(')');
            while(ind!=-1){
                temp.add(st.substring(0,ind)+"()"+st.substring(ind));
                ind = st.indexOf(')',ind+1); 
            }
        }
        System.out.println(n+" : "+temp);
        return helper(n+1,temp,give);
    }
}

File: 22-generate-parentheses/generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        lt = ['()']

        def helper(lt):
            temp = set()
            for i in lt:
                temp.add('('+i+')')
                temp.add(i+'()')
                temp.add('()'+i)
                ind = i.index(')')
                while ind!= -1:
                    temp.add(i[:ind]+'()'+i[ind:])
                    try:
                        ind = i.index(')',ind+1)
                    except ValueError:
                        break
            
            return list(temp)



        for _ in range(1,n):
            lt = helper(lt)
            print(lt)
        
        return lt

        